File: programmers/BFSDFS_네트워크_210416.py
visited = [0 for i in range(200)]

def bfs(start, n, computers):
    queue = []
    queue.append(start)
    
    while queue:
        curr = queue.pop()
        visited[curr] = 1
        for i in range(0, len(computers[curr])):
            if computers[curr][i] == 1 and visited[i] == 0:
                queue.append(i)

def dfs(start, n, computers):
    visited[start] = 1
    for i in range(0, n):
        if computers[start][i] == 1 and visited[i] == 0:
            visited[i] = 1
            dfs(i, n, computers)
    

def solution(n, computers):
    answer = 0
    
    for i in range(0, n):
        if visited[i] == 0:
            #bfs(i, n, computers)
            dfs(i, n, computers)
            answer += 1
    
    return answer

File: programmers/유형별 문제/BFSDFS_네트워크_210416.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>

using namespace std;

int visited[200];
void BFS(int start, int n, vector<vector<int>>& board);
void DFS(int start, int n, vector<vector<int>>& board);

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    for (int i = 0; i < n; ++i) {
        if (visited[i] == 0) {
	    //BFS(i, n, computers);
	    DFS(i, n, computers);
            answer += 1;
        }
    }
    return answer;
}

void BFS(int start, int n, vector<vector<int>>& board) {
    queue<int> q;
    q.push(start);

    while (!q.empty()) {
        int curr = q.front();
        q.pop();
        visited[curr] = 1;
        for (int i = 0; i < n; ++i) {
            if (visited[i] == 0 && board[curr][i] == 1) {
                q.push(i);
            }
        }
    }

    return;
}

void DFS(int start, int n, vector<vector<int>> &board) {
    visited[start] = 1;
    for (int i = 0; i < n; i++) {
        if (board[start][i] == 1 && visited[i] == 0) {
            visited[i] = 1;
            DFS(i, n, board);
        }
    }
    return;
}
